### PR TITLE
Show remove overdue appointment reasons in a separate screen

### PR DESCRIPTION
**Pull request by msasikanth**

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientBottomSheet.kt [hunk 1]
```diff
@@ -162,6 +162,8 @@ class ContactPatientBottomSheet : BaseBottomSheet<
       val selectedDate = result.result as SelectedDate
       val event = ManualDateSelected(selectedDate = selectedDate.date, currentDate = LocalDate.now(userClock))
       hotEvents.onNext(event)
+    } else if (requestType == RemoveOverdueAppointmentResult && result is Succeeded) {
+      router.pop()
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientBottomSheet.kt:166]: If a single screen can navigate to multiple screens for results, then the individual request types should be a `sealed class` type so that the result handling is enforced. Since this screen has at least two result types, this should be followed here.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,7 +8,7 @@
 - [In Progress: 08 Apr 2021] Add automatic performance profiling for the Room database queries
 
 ### Changes
-- [In Progress: 12 Apr 2021] Show remove overdue appointment reasons in a separate screen
+- Show remove overdue appointment reasons in a separate screen
```
Comment (vinaysshenoy) [inline CHANGELOG.md:11]: Technically, this should be a separate commit.

**Review by vinaysshenoy — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (msasikanth) [reply CHANGELOG.md:11]: Ah, I think I accidentally combine it with the last commit. My bad 😅

File: app/src/main/java/org/simple/clinic/contactpatient/views/RemoveAppointmentReasonItem.kt [hunk 3]
```diff
@@ -19,23 +21,25 @@ data class RemoveAppointmentReasonItem(
...
+  override fun layoutResId(): Int = R.layout.removeappointment_reasonitem
 
   override fun render(holder: BindingViewHolder, subject: Subject<Event>) {
-    val binding = holder.binding as ContactpatientRemoveappointmentReasonitemBinding
+    val binding = holder.binding as RemoveappointmentReasonitemBinding
 
     binding.removalReasonButton.setText(reason.displayText)
     binding.removalReasonButton.isChecked = isSelected
     binding.removalReasonButton.setOnClickListener { subject.onNext(Event.Clicked(reason)) }
+    binding.divider.visibleOrGone(showDivider)
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/views/RemoveAppointmentReasonItem.kt:42]: `DividerItemDecorator` was causing UI glitches whenever the list gets updated (the last divider is blinking every time the item selection is changed). To avoid that issue I went with having a divider inside the item.

File: app/src/main/java/org/simple/clinic/removeoverdueappointment/RemoveOverdueAppointmentScreen.kt [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        .ofType<RemoveAppointmentReasonItem.Event.Clicked>()
+        .map { it.reason }
+        .map(::RemoveAppointmentReasonSelected)
+  }
+
+  private fun toolbarNavigationClicks(): Observable<RemoveOverdueEvent> {
+    return toolbar
+        .navigationClicks()
+        .map { CloseClicked }
+  }
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/removeoverdueappointment/RemoveOverdueAppointmentScreen.kt:123]: Yeah, I usually follow that approach but I felt like we are not capturing that flow in tests. So that's why I ended up making it part of the loop for this feature. WDYT?

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientBottomSheet.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientBottomSheet.kt:166]: Yeah, good point.

[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientBottomSheet.kt:166]: But in this scenario, we cannot make it as a sealed class because we are getting the request type for date picker from the key factory. So `RemoveOverdueAppointmentResult` is the only request type in this screen.

File: app/src/main/java/org/simple/clinic/removeoverdueappointment/RemoveOverdueAppointmentScreen.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/removeoverdueappointment/RemoveOverdueAppointmentScreen.kt:123]: I only prefer to capture things in tests that add value. A purely navigation related functionality doesn't add too much value, but a considerable amount of maintenance burden, so I prefer not to make these part of the loop.

File: app/src/main/java/org/simple/clinic/contactpatient/views/RemoveAppointmentReasonItem.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/contactpatient/views/RemoveAppointmentReasonItem.kt:42]: Hmm, this seems alright for now then. But maybe it's something we should revisit and see. A glitching divider is relatively minor compared to the complexity involved in maintaining this, IMO.

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientBottomSheet.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientBottomSheet.kt]: Ah, I see. We could still lift it out and use the `when` clause, right? It won't be exhaustive, but might still be easier to read.

[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientBottomSheet.kt]: Yep, I have done that.

File: app/src/main/java/org/simple/clinic/contactpatient/views/RemoveAppointmentReasonItem.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/views/RemoveAppointmentReasonItem.kt:42]: Yeah, need to revisit the `DividerItemDecorator`. Since that involves playing canvas I didn't want to do it in this PR. Will open a different PR addressing those issues.

File: app/src/main/java/org/simple/clinic/removeoverdueappointment/RemoveOverdueAppointmentScreen.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/removeoverdueappointment/RemoveOverdueAppointmentScreen.kt:123]: Alright, let me remove that event and push it.

**Review by vinaysshenoy — APPROVED** [review]
